packages/jest-mock/src/__tests__/index.test.ts

```diff
@@ -1089,6 +1098,83 @@ describe('moduleMocker', () => {
     });
   });
 
+  describe('mockWhen', () => {
+    it('mocks specific args', () => {
+      const fn = moduleMocker.fn();
+      fn.mockReturnValue('generic output');
+      fn.mockWhen('special input').mockReturnValue('special output');
+
+      expect(fn('arbitrary')).toBe('generic output');
+      expect(fn('special input')).toBe('special output');
+    });
+
+    it('works with several whens', () => {
+      const fn = moduleMocker.fn();
+      fn.mockImplementation(num => `${num}teen`);
+      fn.mockWhen('one').mockReturnValue('eleven');
+      fn.mockWhen('two').mockReturnValue('twelve');
+      fn.mockWhen('three').mockReturnValue('thirteen');
+
+      expect(fn('one')).toBe('eleven');
+      expect(fn('two')).toBe('twelve');
+      expect(fn('three')).toBe('thirteen');
+      expect(fn('four')).toBe('fourteen');
+    });
+
+    it('works without a default implementation', () => {
+      const fn = moduleMocker.fn();
+      fn.mockWhen('special input').mockReturnValue('special output');
+
+      expect(fn('arbitrary')).toBeUndefined();
+      expect(fn('special input')).toBe('special output');
+    });
+
+    it('supports matchers', () => {
+      const fn = moduleMocker.fn();
+      fn.mockReturnValue('hello');
+      fn.mockWhen(expect.any(String)).mockImplementation(v => `hello, ${v}`);
+      fn.mockWhen('hello', expect.any(String)).mockImplementation(
+        (_, y) => `morning, ${y}`,
+      );
+
+      expect(fn(3)).toBe('hello');
+      expect(fn('world')).toBe('hello, world');
+      expect(fn('hello', 'friend')).toBe('morning, friend');
+    });
+
+    it('still tracks all calls on the root mock', () => {
+      const fn = moduleMocker.fn();
+      fn.mockReturnValue('generic output');
+      fn.mockWhen('special input').mockReturnValue('special output');
+
+      expect(fn('arbitrary')).toBe('generic output');
+      expect(fn('special input')).toBe('special output');
+      expect(fn).toHaveBeenNthCalledWith(1, 'arbitrary');
+      expect(fn).toHaveBeenNthCalledWith(2, 'special input');
+      expect(fn).toHaveBeenCalledTimes(2);
+    });
+
+    it('supports spies', () => {
+      class TestClass {
+        public calls: Record<string, number> = {};
+        increment(arg: string) {
+          this.calls[arg] = (this.calls[arg] ?? 0) + 1;
+          return this.calls[arg];
+        }
+      }
+
+      const instance = new TestClass();
+      const spy = spyOn(instance, 'increment');
+      spy.mockWhen('hello').mockReturnValue(-3);
+
+      expect(instance.increment('goodbye')).toBe(1);
+      expect(instance.increment('goodbye')).toBe(2);
+      expect(instance.increment('hello')).toBe(-3);
+      expect(instance.increment('hello')).toBe(-3);
+      expect(instance.increment('goodbye')).toBe(3);
+    });
+  });
```

Comment:
Very exciting! Thank you for working on this. I hope you don't mind I have some thoughts and questions: 
1. What happens when a mock is reset? Do the when's get cleared?
2. What does the failure console output look like? It does it say something like "Expected to be called with X but was called with Y?"
3. What happens if you call with less than or more than the arguments defined in the mockWhen? I assume it's not a match and undefined is returned?
4. Any ideas on how to support a ["default behavior"](https://github.com/timkindberg/jest-when#supports-default-behavior)? I had to add it to jest-when because ppl were defining mocks in a beforeEach and then wanted to define a "default" behavior in each test... so defaultReturnValue (et al) can be set AFTER any `.whenCalledWith` have been set up and it will still be the "default" behavior. For example with your implementation, could we do `jest.fn().whenCalledWith('foo').mockReturnValue(true).mockReturnValue(false)` and have the second mockReturnValue be the "default"? Or would it override the first mockReturnValue?
5. Any idea how we might support [function matcher](https://github.com/timkindberg/jest-when#supports-function-matchers)? Not necessarily now, but maybe eventually.
6. Any ideas how we might support something like the [`when.allArgs` feature](https://github.com/timkindberg/jest-when#supports-matching-or-asserting-against-all-of-the-arguments-together-using-whenallargs) from jest-when? Again, not now, but eventually. 

Points 5 and 6 are a bit more powerful features but they do unlock the full flexibility that users wanted. Those features came later in jest-when's API though. Especially the allArgs feature, check out some use cases: https://github.com/timkindberg/jest-when/blob/master/src/when.test.js#L268-L339.

Reply:
Thanks for the thoughts!

> 1. What happens when a mock is reset? Do the when's get cleared?

Yep, they get cleared, just like `mockImplementation`s etc. (Added a test for this.)

> 2. What does the failure console output look like? It does it say something like "Expected to be called with X but was called with Y?"

There's no failure at all: if the when doesn't match then you get whatever other behavior you've configured (or `undefined` if you didn't). If you want to say this function was called only with certain arguments, there are already ways to do that (`expect.toHaveBeenCalledWith`). With this API I want to support the case where you might now (or in the future) call the mock with several sets of arguments, and only care about some of those. (For example, at my job we have `enqueueTask(taskType, ...)`, and a lot of tests want to mock it out, but then do something special for certain task types of interest.) Of course if you want that behavior you can set the default to be to throw.

> 3. What happens if you call with less than or more than the arguments defined in the mockWhen? I assume it's not a match and undefined is returned?

Good question, I had to try this to find out. (And I've added a test case.) It's not a match, unless the difference is a present/absent `undefined`. (So `whenCalledWith(1)` will not match `fn(1, 1)`, but will match `fn(1, undefined)`.) In principle we could require the number of arguments match exactly but I think this is probably the least surprising behavior?

> 4. Any ideas on how to support a ["default behavior"](https://github.com/timkindberg/jest-when#supports-default-behavior)? I had to add it to jest-when because ppl were defining mocks in a beforeEach and then wanted to define a "default" behavior in each test... so defaultReturnValue (et al) can be set AFTER any `.whenCalledWith` have been set up and it will still be the "default" behavior. For example with your implementation, could we do `jest.fn().whenCalledWith('foo').mockReturnValue(true).mockReturnValue(false)` and have the second mockReturnValue be the "default"? Or would it override the first mockReturnValue?

Interesting question. You can as you probably saw already do
```ts
fn.mockReturnValue(defaultValue)
fn.whenCalledWith(1).mockReturnValue(specialValue)
```
But they have to be in that order; if you do the opposite order you just always get the default. We could change that -- say `whenCalledWith` always takes precedence? I feel like that would be a little confusing since `whenCalledWith(expect.any(number))` doesn't take precedence over `whenCalledWith(1)` (and can't, really, in general -- you don't always have a total ordering), but it would support the case you describe.

Of course if callers really want, they can get the behavior you describe with the following, it's just a bit more awkward.
```ts
let defaultValue = defaultDefaultValue
fn.mockReturnValue(defaultValue)
fn.whenCalledWith(1).mockReturnValue(specialValue)
defaultValue = otherDefaultValue
```

> 5. Any idea how we might support [function matcher](https://github.com/timkindberg/jest-when#supports-function-matchers)? Not necessarily now, but maybe eventually.

Maybe add support for `expect.satisfies(myPredicate)` to the [asymmetric matchers](https://jestjs.io/docs/expect#asymmetric-matchers)? I feel like that would be generally useful. (Or of course callers can do that themselves via `expect.extend` if they have a common predicate they want to use.)

> 6. Any ideas how we might support something like the [`when.allArgs` feature](https://github.com/timkindberg/jest-when#supports-matching-or-asserting-against-all-of-the-arguments-together-using-whenallargs) from jest-when? Again, not now, but eventually.

Interesting question. This does seem like a useful thing to support eventually.

I want the answer to be `fn.whenCalledWith(...expect.arrayOf(expect.any(Number)))`. Right now there's no `expect.arrayOf` but that's easy to add (or for users to add) if it proves useful. The bigger problem is that you can't actually spread a matcher. In principle we could do something magical to make it work (i.e. have matchers have an iterator that returns a sentinel value, and then if `equals` sees an array with that value, it "expands" it for the purposes of the comparison); that could get a bit confusing and/or complicated though.

Or if we're ok with a special API, it would be easy enough to later implement `fn.whenCalledWithAll(expect.any(Number))` or `fn.whenCalledWith(jest.allArgs(...))`.

Reply:
For the default thing (which I anticipate devs asking for at some point like they did with jest-when) maybe we can add a `whenCalledWithDefault` or something. In jest-when I added `defaultReturnValue` and `defaultResolvedValue` etc which is a lot of API surface to add and document. So having just one more `whenCalledWith` function instead is easier. e.g. 

```ts
fn.whenCalledWith(1).mockReturnValue(specialValue)
fn.defaultCalledWith.mockReturnValue(defaultValue)
```

This seems like it might not be so hard to add but I'm not sure.

Thanks for all your other responses, everything looks good.